CppCoreGuidelines.md

```diff
@@ -6081,42 +6086,45 @@ If `x = x` changes the value of `x`, people will be surprised and bad errors may
 
     Foo& Foo::operator=(Foo&& a) noexcept  // OK, but there is a cost
     {
-        if (this == &a) return *this;  // this line is redundant
-        s = std::move(a.s);
-        i = a.i;
+        if (this != &a) {
+            s = std::move(a.s);
+            i = a.i;
+        }
         return *this;
     }
 
-The one-in-a-million argument against `if (this == &a) return *this;` tests from the discussion of [self-assignment](#Rc-copy-self) is even more relevant for self-move.
-
-##### Note
-
-There is no known general way of avoiding an `if (this == &a) return *this;` test for a move assignment and still get a correct answer (i.e., after `x = x` the value of `x` is unchanged).
+Here, the test `(this != &a)` is not redundant, because the standard library does not guarantee the result
```

Comment:
library guarantees the result is valid-but-unspecified (as noted in the next paragraph you're deleting), even though it wasn't specified well in the past, see http://wg21.link/lwg2468 and http://wg21.link/lwg2839

Reply:
In practice, both libc++ and libstdc++ will set `s` to the empty string. So, if you want `foo = std::move(foo)` to be a no-op, you must either write special-case code to handle the self-assignment case, or use the copy-and-swap idiom.

https://godbolt.org/z/Bsst2o

Perhaps the Guidelines should go further, and explain that if you follow the Rule of Zero (as in the godbolt above), you _will not_ be following rule C.65, because `x = std::move(x)` will not be a no-op. The defaulted move assignment operator will not use copy-and-swap; it will use `std::string`'s move-assignment operator, which is not a no-op on self-move.

The test in this example is _absolutely not redundant_. https://godbolt.org/z/NjL5VR

Reply:
Wait, I didn't realize that you're interpreting C.65 so that `foo = std::move(foo)` must be a no-op. That's definitely wrong if it gives that impression.

Reply:
...actually it seems to imply it with "x is unchanged", even though it is not supported by C.65's own motivation to make self-swap work. This may be worth a new issue.

Reply:
Yep. Right now, rule C.64 says "A move operation should move and leave its source in a valid state" (but potentially *any* valid state), which is the STL's own rule.

Rule C.65 specifically goes stronger; C.65 says "Make move assignment safe for *self-assignment* ... If `x = x` changes the value of `x`, people will be surprised ... after `x = x` the value of `x` is unchanged ... The ISO standard guarantees only a “valid but unspecified” state [but] The rule here is more caution and *insists on complete safety*."

Reply:
right, and this means the mention of self-swap in the lead is irrelevant, the title is misleading (should be "no-op" rather than "safe"), and that the downsides that you're correctly bringing up were omitted.